flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1014 @@
+		 * @throws IOException
+		 */
+		public void overwriteRecordAt(long pointer, DataInputView input, int size) throws IOException {
+			setWritePosition(pointer);
+			outView.write(input, size);
+		}
+
+		/**
+		 * Appends a pointer and a record. The record is read from a DataInputView (this will be the staging area).
+		 * @param pointer The pointer to write (Note: this is NOT the position to write to!)
+		 * @param input The DataInputView to read the record from
+		 * @param recordSize The size of the record
+		 * @return A pointer to the written data
+		 * @throws IOException (EOFException specifically, if memory ran out)
+		 */
+		public long appendPointerAndCopyRecord(long pointer, DataInputView input, int recordSize) throws IOException {
+			setWritePosition(appendPosition);
+			final long oldLastPosition = appendPosition;
+			outView.writeLong(pointer);
+			outView.write(input, recordSize);
+			appendPosition += 8 + recordSize;
+			return oldLastPosition;
+		}
+
+		/**
+		 * Appends a pointer and a record.
+		 * @param pointer The pointer to write (Note: this is NOT the position to write to!)
+		 * @param record The record to write
+		 * @return A pointer to the written data
+		 * @throws IOException (EOFException specifically, if memory ran out)
+		 */
+		public long appendPointerAndRecord(long pointer, T record) throws IOException {
+			setWritePosition(appendPosition);
+			return noSeekAppendPointerAndRecord(pointer, record);
+		}
+
+		/**
+		 * Appends a pointer and a record. Call this function only if the write position is at the end!
+		 * @param pointer The pointer to write (Note: this is NOT the position to write to!)
+		 * @param record The record to write
+		 * @return A pointer to the written data
+		 * @throws IOException (EOFException specifically, if memory ran out)
+		 */
+		public long noSeekAppendPointerAndRecord(long pointer, T record) throws IOException {
+			final long oldLastPosition = appendPosition;
+			final long oldPositionInSegment = outView.getCurrentPositionInSegment();
+			final long oldSegmentIndex = outView.currentSegmentIndex;
+			outView.writeLong(pointer);
+			buildSideSerializer.serialize(record, outView);
+			appendPosition += outView.getCurrentPositionInSegment() - oldPositionInSegment +
+				outView.getSegmentSize() * (outView.currentSegmentIndex - oldSegmentIndex);
+			return oldLastPosition;
+		}
+
+		public long getAppendPosition() {
+			return appendPosition;
+		}
+
+		// ----------------------- Input -----------------------
+
+		public void setReadPosition(long position) {
+			inView.setReadPosition(position);
+		}
+
+		public long getReadPosition() {
+			return inView.getReadPosition();
+		}
+
+		public long readLong() throws IOException {
+			return inView.readLong();
+		}
+
+		public T readRecord(T reuse) throws IOException {
+			return buildSideSerializer.deserialize(reuse, inView);
+		}
+	}
+
+
+	private final class StagingOutputView extends AbstractPagedOutputView {
```

Comment:
I'm thinking about this one, but it looks problematic, because `SimpleCollectingOutputView.nextSegment` always requests a new segment from the `memorySource`, but `StagingOutputView` reuses its segments after a reset.

`SimpleCollectingOutputView.reset` has a similar problem: it works only if there are no segments at the moment.

So I think I would have to override `nextSegment`, and `reset`, which defeats the purpose of simplifying things.